internal/dao/secret.go

```diff
@@ -0,0 +1,94 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of K9s
+
+package dao
+
+import (
+	"fmt"
+	"strings"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// Secret represents a secret K8s resource.
+type Secret struct {
+	Resource
+	decode bool
+}
+
+// Describe describes a secret that can be encoded or decoded.
+func (s *Secret) Describe(path string) (string, error) {
+	encodedDescription, err := s.Resource.Describe(path)
+
+	if err != nil {
+		return "", err
+	}
+
+	if !s.decode {
+		return encodedDescription, nil
+	}
+
+	return s.Decode(encodedDescription, path)
+}
+
+// SetDecode sets the decode flag.
+func (s *Secret) SetDecode(flag bool) {
+	s.decode = flag
+}
+
+// Decode removes the ecnode part from the secret's description and appends the
+// secret's decoded data.
+func (s *Secret) Decode(encodedDescription, path string) (string, error) {
+	o, err := s.GetFactory().Get(s.GVR(), path, true, labels.Everything())
+
+	if err != nil {
+		return "", err
+	}
+
+	dataEndIndex := strings.Index(encodedDescription, "====") + 4
+
+	if dataEndIndex == -1 {
+		return "", fmt.Errorf("unable to find data section in secret description")
+	}
+
+	d, err := ExtractSecrets(o.(*unstructured.Unstructured))
+
+	if err != nil {
+		return "", err
+	}
+
+	// Remove the encoded part from k8s's describe API
+	// More details about the reasoning of index: https://github.com/kubernetes/kubectl/blob/v0.29.0/pkg/describe/describe.go#L2542
+	decodedDescription := string([]rune(encodedDescription)[0:dataEndIndex])
+
+	for k, v := range d {
+		decodedDescription = fmt.Sprintf("%s\n%s:\t%s", decodedDescription, k, v)
+	}
+
+	return decodedDescription, nil
+}
+
+// ExtractSecrets takes an unstructured object and attempts to convert it into a
+// Kubernetes Secret.
+// It returns a map where the keys are the secret data keys and the values are
+// the corresponding secret data values.
+// If the conversion fails, it returns an error.
+func ExtractSecrets(o *unstructured.Unstructured) (map[string]string, error) {
+	var secret v1.Secret
```

Comment:
yes, I could've done `o.Object["data"]`, but I did not see any place to do that, so I followed the pattern of using `runtime.DefaultUnstructuredConverter.FromUnstructured`
Is it any issue regarding how it is now?

Reply:
Hydrating a full object from json can be expensive, but not a big deal since this is a one off. 
Was thinking since we have the map we could just retrieve the data section directly via jsonpath?

Reply:
actually, i have just checked now and without `runtime.DefaultUnstructuredConverter.FromUnstructured` I get only the base64 value

Reply:
Ah! right.